### ISSUE TITLE
[Bug] Poor performance when processing large amounts of LLM text data

### ISSUE THREAD

**Issue by Cathy0908** [ray-project/ray#45928]
### What happened + What you expected to happen

I want to use ray to do large-scale text data cleaning tasks, and extracted 5 million data from the open source redpajama github dataset for testing, which is about **58G**. I used two 128 vCPU, 256 GiB nodes for testing, which took **more than 2 hours**. 

However, I used a 128 vCPU, 256 GiB node, did not use the ray framework, and used the `datasets` api to do multi-process processing, which only took about **20 minutes**.

When using ray, it always reports an error: (raylet) A worker died or was killed while executing a task by an unexpected system error. It should be that the worker was killed continuously due to out-of-memory. I have increased --object-store-memory, reduced concurrency and other debugging settings, but still cannot improve.

What should I do to improve performance. Now the performance of ray 2 nodes and single machines without ray is very different.

![image](https://github.com/ray-project/ray/assets/30484308/f2094fa0-ef71-4424-a1ca-0e3e027c11a4)

### Versions / Dependencies

- ray: 2.10.0
- python: 3.8.15

### Reproduction script

ray script: 

```python

import ftfy
import time


def FixUnicodeMapper(sample):
    text_key = 'content'
    normalization = 'NFKC'
    sample[text_key] = ftfy.fix_text(sample[text_key], normalization=normalization)
    return sample


if __name__ == "__main__":
    import ray
    from ray.data import read_json

    ray.init()

    num_nodes = ray.nodes()
    num_nodes_used = len(num_nodes)
    print(">>> Number nodes:", num_nodes_used)

    data_path = "./datasets/redpajama_github/json_parts_500w/"   # 224 jsonl files
    export_path = './out_dir/'

    s_time = time.time()

    dataset = read_json(data_path)
    res_datasets = dataset.map(FixUnicodeMapper)
    res_datasets.write_json(export_path, force_ascii=False)

    print('cost time: ', time.time() - s_time)

    ray.shutdown()

```

### Issue Severity

None

**Comment by Bye-legumes** [comment]
Hi,  you should try to use [map_batches()](https://docs.ray.io/en/latest/data/api/doc/ray.data.Dataset.map_batches.html#ray.data.Dataset.map_batches) to process vectorized data. Map is not efficient as it will process each row and create a task...

 Maybe you can change number of parallels by setting `concurrency` . also based you the screenshot, the object memory is not full but it's OOM, this is typically because there are not enough memory for work heap, i.e. 30% store will results in 70% heap while, 60 % store results in 40% heap. So what you need is to reduce the object memory store so that the work heap can take more memory.

**Comment by Cathy0908** [comment]
Thanks very much, I'll try it.

**Comment by anyscalesam** [comment]
If your setup allows as well please grab the latest version of Ray; there was a bunch of perf and stability updates we did from 2.10 to 2.24, which is latest. (we've started releasing Ray weekly to keep up with the rapid pace of change in the AI/ML space in recent months)

**Comment by Cathy0908** [comment]
@Bye-legumes @anyscalesam 

Thanks. According to the previous suggestions, I tried to use map_batches, and adjust the batch_size, concurrency, and memory parameters, and the performance has improved. But it still does not meet expectations. Currently, it takes about 1 hour for 2 nodes, which is still a bit different from the half-hour for a single machine. But I am using 2.10. I will continue to test it with the latest version. Thanks

**Comment by Cathy0908** [comment]
I used version 2.24, but the performance is still not improved. Is there any way to speed up?

**Comment by Bye-legumes** [comment]
Hi, your plot is very useful! Thanks @Cathy0908 
Can you make it into a actor instead of task so it will not release the resource after task finished to reduce the task init time.
After the test above if the cpu usage is still not very high, can you change the resource requirements ,i.e. cpu=0.5 or 0.25 so it will have higher parallel(concurrency).

**Comment by Cathy0908** [comment]
@Bye-legumes Thank you for your suggestions. I changed to actor and tried several sets of parameters, but it didn't seem to work. Later I tried to increase the machine's memory to 512G. It was obviously improved, the performance has reached my expectations, and it does not kill workers frequently. 

However, it seems that when the memory reaches a certain proportion, the performance seems not very good , unless there is enough memory

```python
import ftfy
import time
import ray
from ray.data import read_json
from ray.data.context import DatasetContext
DatasetContext.max_errored_blocks = 2


class FixUnicodeMapperActor:
    def __init__(self):
        self.text_key = 'content'
        self.normalization = 'NFKC'

    def __call__(self, samples):
        normalize_text = list(map(
            lambda t: ftfy.fix_text(t, normalization=self.normalization), samples[self.text_key]))
        samples[self.text_key] = normalize_text
        return samples


if __name__ == "__main__":
    ray.init()

    num_nodes = ray.nodes()
    print(">>>>>> Number nodes:", len(num_nodes))
    
    data_path = "/mnt/data/json_parts_500w_v4/"
    export_path = '/mnt/data/outputs/actor_v4/'

    s_time = time.time()

    dataset = read_json(data_path)
    
    res_datasets = dataset.map_batches(
        FixUnicodeMapperActor, 
        batch_size=1024*2,
      	num_cpus=0.25,
        concurrency=1024
    )

    res_datasets.write_json(export_path, force_ascii=False)
    
    print('cost time: ', time.time() - s_time)
    ray.shutdown()

```python

**Comment by scottjlee** [comment]
>Later I tried to increase the machine's memory to 512G. It was obviously improved

Great to hear! So based on the above discussion / screenshots, it looks like the main issue is that Ray is not dedicating enough memory to the Ray Data tasks. @Cathy0908 What commands were you using to adjust the heap memory / object store capacity / etc?

**Comment by Superskyyy** [comment]
@scottjlee I think the main friction here is that currently there's no way to adjust heap and preallocated object store ratio dynamically, which in many cases user simply goes by the default value and it just sits there taking up the available heap. 
@Bye-legumes and I are thinking of adding such a capability to the plasma store, so that it could be scaled elastically to make room for heap (also scale up), and such a poor performance can be automatically mitigated in many use cases. What do you think?

**Comment by scottjlee** [comment]
Yes, that sounds like it could work. Please keep us updated with a PR, or if you have any questions. Thanks!

**Comment by anyscalesam** [comment]
@scottjlee can you assign a priority please?

**Comment by richardliaw** [comment]
closing this out since stale; feel free to open a new issue if this is a still an issue
